### PR TITLE
Rich sift describer

### PR DESCRIPTION
**Pull request by shauidu**
add a rich_sift_describer, to support computing given region's decriber.

### CODE REVIEW

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    regions_type->Descriptors().clear();
+    regions_type->Descriptors().reserve(2000);
+
+    int curr_feature_index = 0;
+    int octave_index = 0;
+    while (true) {
+      // Update gradient before launching parallel extraction
+      vl_sift_update_gradient(filt);
+
+      while (curr_feature_index < features.size() &&
```
Comment (chubei-oppen) [inline src/nonFree/rich_sift/Rich_SIFT_describer.hpp]: use openmp here to accelerate computation

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+  }
+
+  /**
+  @brief Detect regions on the image and compute their attributes (description)
+  @param image Image.
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  @return regions The detected regions and attributes (the caller must delete the allocated data)
+  */
+  std::unique_ptr<Regions> Describe(
```
Comment (chubei-oppen) [inline src/nonFree/rich_sift/Rich_SIFT_describer.hpp:125]: test this implementation by recomputing all features on image

File: src/openMVG/features/feature.cpp [hunk 3]
```diff
@@ -81,6 +81,69 @@ std::istream& operator>>(std::istream& in, SIOPointFeature& obj)
...
+         (s_ == b.s_);
+};
+
+bool Rich_SIOPointFeature::operator !=(const Rich_SIOPointFeature& b) const {
+  return !((*this)==b);
+};
+
+std::ostream& operator<<(std::ostream& out, const Rich_SIOPointFeature& obj)
+{
+  const PointFeature *pf = static_cast<const PointFeature*>(&obj);
```
Comment (chubei-oppen) [inline src/openMVG/features/feature.cpp]: cast to its direct base class SIOPointFeature

File: src/openMVG/features/feature.cpp [hunk 4]
```diff
@@ -81,6 +81,69 @@ std::istream& operator>>(std::istream& in, SIOPointFeature& obj)
...
+  return out << *pf << " " << obj.scale_ << " " << obj.orientation_ << " " << obj.octave_ <<
+         " " << obj.ix_ << " " << obj.iy_ << " " << obj.is_ << " " << obj.s_;
+}
+
+std::istream& operator>>(std::istream& in, Rich_SIOPointFeature& obj)
+{
+  PointFeature *pf = static_cast<PointFeature*>(&obj);
```
Comment (chubei-oppen) [inline src/openMVG/features/feature.cpp]: use SIOPointFeature here too

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp (hunk 1 above)
Comment (shauidu) [reply src/nonFree/rich_sift/Rich_SIFT_describer.hpp]: done

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp (hunk 2 above)
Comment (shauidu) [reply src/nonFree/rich_sift/Rich_SIFT_describer.hpp:125]: the origin description is calculated using a double orientation, then SIOPointFeature serialize the orientation as a float, so the recalculated description is different.
I convert the orientation to float at the very beginning to solve this.

File: src/openMVG/features/feature.cpp (hunk 3 above)
Comment (shauidu) [reply src/openMVG/features/feature.cpp]: done

File: src/openMVG/features/feature.cpp (hunk 4 above)
Comment (shauidu) [reply src/openMVG/features/feature.cpp]: done

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    const image::Image<unsigned char>* mask = nullptr
+  ) override
+  {
+    return DescribeSIFT(image, mask);
+  }
+
+  /**
+  @brief compute given regions' attributes (description)
+  @param image Image.
+  @param mask 8-bit gray image for keypoint filtering (optional).
```
Comment (chubei-oppen) [inline src/nonFree/rich_sift/Rich_SIFT_describer.hpp]: remove unused parameter comments

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    regions_type->Descriptors().resize(features.size());
+
+    int curr_feature_index = 0;
+    int octave_index = 0;
+    while (true) {
+      // Update gradient before launching parallel extraction
+      vl_sift_update_gradient(filt);
+
+      int begin = curr_feature_index;
+      int& end = curr_feature_index;
```
Comment (chubei-oppen) [inline src/nonFree/rich_sift/Rich_SIFT_describer.hpp]: a little confusing. just use curr_feature_index

File: src/nonFree/rich_sift/Rich_SIFT_describer.hpp [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+        key.is = f.is();
+        key.x = f.x();
+        key.y = f.y();
+        key.s = f.s();
+        key.sigma = f.scale();
+
+        vl_sift_calc_keypoint_descriptor(filt, &descr[0], &key, f.orientation());
+
+        siftDescToUChar(&descr[0], descriptor, _params._root_sift);
+        #ifdef OPENMVG_USE_OPENMP
```
Comment (chubei-oppen) [inline src/nonFree/rich_sift/Rich_SIFT_describer.hpp]: not necessarily a critical region

**Pull request merged**